omp/base/kernel_launch.hpp

```diff
@@ -0,0 +1,304 @@
+    {
+        return {param.data, num_cols};
+    }
+};
+
+
+template <typename T>
+typename device_unpack_2d_impl<typename to_device_type_impl<T>::type>::type
+map_unpack_to_device(T &&param, size_type num_cols)
+{
+    return device_unpack_2d_impl<typename to_device_type_impl<T>::type>::unpack(
+        to_device_type_impl<T>::map_to_device(param), num_cols);
+}
+
+
+}  // namespace omp
+}  // namespace kernels
+
+
+template <typename ValueType>
+kernels::omp::compact_dense_wrapper<ValueType> compact(
+    matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {mtx->get_values()};
+}
+
+
+template <typename ValueType>
+kernels::omp::compact_dense_wrapper<const ValueType> compact(
+    const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {mtx->get_const_values()};
+}
+
+
+template <typename ValueType>
+ValueType *vector(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return mtx->get_values();
+}
+
+
+template <typename ValueType>
+const ValueType *vector(const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return mtx->get_const_values();
+}
+
+
+template <typename KernelFunction, typename... KernelArgs>
+void run_kernel(std::shared_ptr<const OmpExecutor> exec, KernelFunction fn,
+                size_type size, KernelArgs &&... args)
+{
+#pragma omp parallel for
+    for (size_type i = 0; i < size; i++) {
+        [&]() { fn(i, kernels::omp::map_to_device(args)...); }();
+    }
+}
+
+template <size_type cols, typename KernelFunction, typename... KernelArgs>
+void run_kernel_fixed_cols(std::shared_ptr<const OmpExecutor> exec,
+                           KernelFunction fn, dim<2> size,
+                           KernelArgs &&... args)
+{
+    const auto rows = size[0];
+#pragma omp parallel for
+    for (size_type row = 0; row < rows; row++) {
+#pragma unroll
+        for (size_type col = 0; col < cols; col++) {
+            [&]() {
+                fn(row, col, kernels::omp::map_unpack_to_device(args, cols)...);
+            }();
+        }
+    }
+}
+
+template <size_type remainder_cols, typename KernelFunction,
+          typename... KernelArgs>
+void run_kernel_blocked_cols(std::shared_ptr<const OmpExecutor> exec,
+                             KernelFunction fn, dim<2> size,
+                             KernelArgs &&... args)
+{
+    const auto rows = size[0];
+    const auto cols = size[1];
+    const auto rounded_cols = cols / 4 * 4;
+    GKO_ASSERT(rounded_cols + remainder_cols == cols);
+#pragma omp parallel for
+    for (size_type row = 0; row < rows; row++) {
+        for (size_type base_col = 0; base_col < rounded_cols; base_col += 4) {
+#pragma unroll
+            for (size_type i = 0; i < 4; i++) {
```

Comment:
Nice! I guess a good aspect of common kernels is that you can now afford to spend time putting all these optimizations in one place and they pay off everywhere they're used. I guess this is to aid interleaving, but does this still vectorize?

Reply:
Exactly, this was one of the ideas behind this approach :) I did not check the generated code in detail, it should definitely be possible to auto-vectorize, but since the kernels are memory-bound, I am not sure how much of a difference vectorization would make here.